### PR TITLE
Test suite test_suite_pk test pk_rsa_overflow passes valid parameters

### LINKED ISSUE #1179: In test_suite_pk.function, the invalid size tests are not testing what they should
- Type: Bug
- Priority: Minor

In `tests/suites/test_suite_pk.function`, there are calls to `mbedtls_pk_sign`, `mbedtls_pk_verify` and `mbedtls_pk_verify_ext` that pass a NULL pointer as the signature buffer and a size that is too large. The test verifies that the function returns `MBEDTLS_ERR_PK_BAD_INPUT_DATA`, but it's actually doing that because of the NULL pointer and not because of the size. Since the job of the test is to validate the size check, the test is not doing its job, it's just passing by coincidence.

The test should pass a valid pointer instead.

Detected in #900

### PR DESCRIPTION
**Pull request by itayzafrir**
Test suite `test_suite_pk` test `pk_rsa_overflow` passing valid parameters for hash/hash_len and sig/sig_len. Resolves #1179 

## Status
**READY**

## Requires Backporting

Yes
Which branch?
mbedtls 2.1
mbedtls 2.7

## Todos
- [ ] Tests
- [ ] Documentation
- [X] Changelog updated
- [X] Backported

### CODE REVIEW

File: tests/suites/test_suite_pk.function [hunk 1]
```diff
@@ -417,26 +417,30 @@ exit:
 void pk_rsa_overflow( )
 {
     mbedtls_pk_context pk;
-    size_t hash_len = SIZE_MAX;
+    size_t hash_len = SIZE_MAX, sig_len = SIZE_MAX;
+    unsigned char hash[50], sig[1000];
```
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_pk.function]: Please use a smaller buffer for `sig`. This is a preexisting issue in other test cases as well, the large sizes cause stack overflows on some platforms, we're fixing them little by little as we modify the code for other reasons.

File: ChangeLog [hunk 2]
```diff
@@ -33,6 +33,8 @@ Bugfix
    * Fix mbedtls_x509_crt_profile_suiteb, which used to reject all certificates
      with flag MBEDTLS_X509_BADCERT_BAD_PK even when the key type was correct.
      In the context of SSL, this resulted in handshake failure. #1351
+   * Fix test suite test_suite_pk test pk_rsa_overflow passes valid parameters 
+     for hash and sig. #1179
```
Comment (gilles-peskine-arm) [inline ChangeLog]: We use complete English sentences in the changelog, and it isn't just `pk_rsa_overflow`. Suggestion: “In test_suite_pk, pass otherwise valid parameters when testing hash length overflow. \#1179”

**Comment by gilles-peskine-arm** [comment]
@itayzafrir Apologies, I'd been looking at a branch with some refactoring. Indeed `pk_rsa_alt` is fine.

**Comment by gilles-peskine-arm** [comment]
Would you mind rebasing your branch so that it has a clean commit history, before others review it?

In general we try to avoid rebasing because that makes the review history hard to follow. But here, since it's a simple fix which could fit in a single commit, a clean commit history would be a benefit. Also, merge commits are hard to review, please avoid merge commits in the PR unless there's a good reason for it, such as a conflict in a file other than `ChangeLog` (Git can't resolve ChangeLog conflicts natively but we have a script to fix them when merging). In general, it's fine to base a PR on top of the previous release, you don't need to track `development` if it doesn't interact with other changes in progress.

**Comment by itayzafrir** [comment]
@gilles-peskine-arm rebased as requested

File: ChangeLog [hunk 3]
```diff
@@ -34,6 +34,9 @@ Bugfix
      with flag MBEDTLS_X509_BADCERT_BAD_PK even when the key type was correct.
      In the context of SSL, this resulted in handshake failure. #1351
    * Fix Windows x64 builds with the included mbedTLS.sln file. #1347
+   * In test_suite_pk pass valid parameters when testing for hash length 
+     overflow. #1179
+   * Fix Windows x64 builds with the included mbedTLS.sln file. #1347
```
Comment (Patater) [inline ChangeLog]: Why does your commit add `Fix Windows x64 builds with the included mbedTLS.sln file. #1347`? I don't think your commit should add this, as it doesn't fix Windows x64 builds.

File: tests/suites/test_suite_pk.function [hunk 4]
```diff
@@ -417,26 +417,30 @@ exit:
...
 {
     mbedtls_pk_context pk;
-    size_t hash_len = SIZE_MAX;
+    size_t hash_len = SIZE_MAX, sig_len = SIZE_MAX;
+    unsigned char hash[50], sig[100];
 
     if( SIZE_MAX <= UINT_MAX )
         return;
 
+    memset( hash, 0x2a, sizeof hash );
```
Comment (Patater) [inline tests/suites/test_suite_pk.function]: It is Mbed TLS style to use parenthesis with `sizeof`, even though it isn't strictly required by C.

See "Extra parentheses for return and sizeof" in the [Mbed TLS Style Guide](https://tls.mbed.org/kb/development/mbedtls-coding-standards).

File: tests/suites/test_suite_pk.function [hunk 5]
```diff
@@ -417,26 +417,30 @@ exit:
 void pk_rsa_overflow( )
 {
     mbedtls_pk_context pk;
-    size_t hash_len = SIZE_MAX;
+    size_t hash_len = SIZE_MAX, sig_len = SIZE_MAX;
+    unsigned char hash[50], sig[100];
```
Comment (Patater) [inline tests/suites/test_suite_pk.function]: Where do the numbers `50` and `100` come from? Add a comment explaining the rationale for their sizes, in case anybody wants to come back later and change them. If the sizes are constant and nobody would ever change them, consider making well named constants for the sizes to improve readability.

File: tests/suites/test_suite_pk.function (hunk 4 above)
Comment (Patater) [reply tests/suites/test_suite_pk.function]: Consider adding a comment explaining why 0x2A is used for initializing the hash.

Comment (gilles-peskine-arm) [reply tests/suites/test_suite_pk.function]: These are preexisting issues, as are the hard-coded buffer sizes. Separate commits to fix these issues would be appreciated, but are outside the scope of the issue that triggered this PR.

Comment (Patater) [reply tests/suites/test_suite_pk.function]: Which issues exactly are preexisting?

At least the `sizeof` issue is freshly added code and should use correct style, even if copy-pasted from a bad example.

Where does 0x2A come from if this is preexisting? It also seems to be freshly added code.

Comment (Patater) [reply tests/suites/test_suite_pk.function]: OK. Unfortunate to have those issues, but I am fine with these as they are consistent with the rest of the file.

File: ChangeLog (hunk 3 above)
Comment (itayzafrir) [reply ChangeLog]: Sorry this was a mistake, the comment 'Fix Windows x64 builds with the included mbedTLS.sln file. #1347' was duplicated by mistake, fixing this.

File: tests/suites/test_suite_pk.function (hunk 5 above)
Comment (itayzafrir) [reply tests/suites/test_suite_pk.function]: These sizes have no true meaning as this test checks for hash length overflows and these buffers are never actually used.

File: tests/suites/test_suite_pk.function (hunk 4 above)
Comment (itayzafrir) [reply tests/suites/test_suite_pk.function]: As @gilles-peskine-arm stated, 0x2a and coding style issues are consistent with the rest of the file.

**Comment by Patater** [comment]
Could you please squash the two commits together? Otherwise, merging this PR will still add commits that introduces noise into the ChangeLog.

Thanks

**Comment by itayzafrir** [comment]
Squashed as @Patater requested
